### PR TITLE
[dockerfile] update requirements and jobs order

### PR DESCRIPTION
**Pull request by pducolin**
What does this PR do?
---------------------

* Use e2e requirements from [buildimages](https://github.com/DataDog/datadog-agent-buildimages/blob/main/requirements/e2e.txt)
* Build the image and publish to a test repo, run integration tests on the test version, push to stable runner on main after tests success

Which scenarios this will impact?
-------------------

None

Motivation
----------

Use e2e python requirements to keep in sync with other agent python requirements, since the image is used by `datadog-agent` pipeline

Run tests on the pipeline image to ensure the image runs fine

Additional Notes
----------------

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -1,14 +1,33 @@
...
+build-runner-image:
+  stage: build
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/docker:20.10.13
+  tags: ["arch:amd64"]
+  variables:
+    RELEASE_IMAGE: "false"
+  rules:
+    - when: on_success
+  script:
+    # Tag with 8 and 12 SHA to match go.mod
```
Comment (vboulineau) [inline .gitlab-ci.yml:23]: Should be removed

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -1,14 +1,33 @@
...
+    - docker buildx build --no-cache --pull --push --label target=build --tag ${CI_REGISTRY_IMAGE_TEST}:${CI_COMMIT_SHORT_SHA} .
```
Comment (vboulineau) [inline .gitlab-ci.yml:24]: I'm not sure the `label target=build` brings any value, image metadata are almost never used.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -31,21 +50,15 @@ integration-testing:
...
-    - TARGET_REPO="${IMAGE_REPOSITORY}/runner"
-    - if [ "$PUSH_IMAGE" = "true" ]; then PUSH_ARG="--push"; fi
-    # Tag with 8 and 12 SHA to match go.mod
-    - docker buildx build --no-cache --pull $PUSH_ARG --tag ${TARGET_REPO}:${CI_COMMIT_SHORT_SHA} --tag ${TARGET_REPO}:${CI_COMMIT_SHA:0:12}  .
-  retry: 2
-
+    - docker pull ${CI_REGISTRY_IMAGE_TEST}:${CI_COMMIT_SHORT_SHA}
+    - docker tag  ${CI_REGISTRY_IMAGE_TEST}:${CI_COMMIT_SHORT_SHA} ${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHORT_SHA} 
+    - docker tag  ${CI_REGISTRY_IMAGE_TEST}:${CI_COMMIT_SHORT_SHA} ${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHA:0:12}
+    - docker push --label target=build ${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHA:0:12} --tag ${IMAGE_REPOSITORY}:${CI_COMMIT_SHA:0:12}  .
```
Comment (vboulineau) [inline .gitlab-ci.yml:61]:
```suggestion
    - docker push ${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHA:0:12}
```

The `docker push` syntax does not work as `docker build`, it only takes a single parameter and does not take any working dir path.

Also I believe we can now remove the `${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHORT_SHA}` tag as it should not be used anywhere anymore.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -31,21 +50,15 @@ integration-testing:
...
+  only:
```
Comment (vboulineau) [inline .gitlab-ci.yml:62]: `only` is deprecated, you should keep the `rules:if` that was already there.

File: .gitlab-ci.yml (hunk 2 above)
Comment (pducolin) [reply .gitlab-ci.yml:24]: It is mentioned in the CI/CD onboarding to Gitlab runners: https://datadoghq.atlassian.net/wiki/spaces/DEVX/pages/2411528655/Onboarding+To+Gitlab+CI#4.-ECR-(Elastic-Container-Registry)-for-a-CI-Image-(optional)

File: .gitlab-ci.yml (hunk 3 above)
Comment (pducolin) [reply .gitlab-ci.yml:61]: Forgot to update it, the label is added at the previous build stage, tags are added with `docker tag`

File: .gitlab-ci.yml (hunk 2 above)
Comment (vboulineau) [reply .gitlab-ci.yml:23]: It's dedicated to `registry.dd.io` not ECR.

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -31,21 +49,14 @@ integration-testing:
...
-  stage: build
+release-runner-image:
+  stage: release
   image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/docker:20.10.13
   tags: ["arch:amd64"]
-  variables:
-    PUSH_IMAGE: "false"
+  script:
+    - docker pull ${CI_REGISTRY_IMAGE_TEST}:${CI_COMMIT_SHORT_SHA}
+    - ${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHA:0:12}
```
Comment (vboulineau) [inline .gitlab-ci.yml:58]:
```suggestion
    - docker tag ${CI_REGISTRY_IMAGE_TEST}:${CI_COMMIT_SHORT_SHA} ${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHA:0:12}
```

**Review by vboulineau — APPROVED** [review]

**Pull request merged**